[13.0] Add module sale_cutoff_time_delivery

grindtildeath opened the pull request:
Depends on:
 - [x] https://github.com/OCA/partner-contact/pull/876

This module allows to define Cutoff times for delivery orders.

If no commitment date is set, any sale order confirmed after the applicable
cutoff time will have its delivery order postponed by one day at the time of
the applicable cutoff, and any sale order confirmed before the applicable
cutoff time will have its delivery order postponed to the time of the applicable
cutoff.

Application of cutoff is defined by partner, with default value being the
cutoff time of the sale order's warehouse.

Cut-off must however be applied on the warehouse by marking the checkbox
"Apply cutoff" so that the computation happens.

grindtildeath committed:
Add module sale_cutoff_time_delivery

rousseldenis commented on sale_cutoff_time_delivery/models/cutoff_time_mixin.py:
```diff
@@ -0,0 +1,31 @@
+        selection="_selection_order_delivery_cutoff_hours", required=True, default="12"
+    )
+    order_delivery_cutoff_minutes = fields.Selection(
+        selection="_selection_order_delivery_cutoff_minutes",
```
Please change to lambda

grindtildeath replied:
AFAICS, this is not needed, as a string arg is called using using `getattr`: https://github.com/odoo/odoo/blob/13.0/odoo/fields.py#L2170

guewen replied:
I think this is what you wanted to link: https://github.com/odoo/odoo/blob/45afa3a20693ebed11ad2fa7ebd1b73772ca5b3d/odoo/fields.py#L2242-L2243 ;)

rousseldenis commented on sale_cutoff_time_delivery/models/cutoff_time_mixin.py:
```diff
@@ -0,0 +1,31 @@
+
+class CutoffTimeMixin(models.AbstractModel):
+
+    _name = "cutoff.time.mixin"
```
Prefer sale.cutoff.time.mixin as this is applicable to sale. Either, externalize in another module

grindtildeath committed:
fixup! Add module sale_cutoff_time_delivery

grindtildeath commented on sale_cutoff_time_delivery/models/cutoff_time_mixin.py:
```diff
@@ -0,0 +1,31 @@
+
+    _name = "sale.cutoff.time.mixin"
+
+    order_delivery_cutoff_hours = fields.Selection(
```
Probably better to harmonize usage with #1101 and use Float

grindtildeath commented on sale_cutoff_time_delivery/readme/ROADMAP.rst:
```diff
@@ -0,0 +1,2 @@
+* Consider timezones
```
This must be fixed as well, although I still don't know how to do it :sweat_smile:

grindtildeath committed:
fixup! fixup! Add module sale_cutoff_time_delivery

grindtildeath committed:
fixup! fixup! fixup! Add module sale_cutoff_time_delivery

grindtildeath committed:
fixup! fixup! fixup! fixup! Add module sale_cutoff_time_delivery

grindtildeath committed:
fixup! fixup! fixup! fixup! fixup! Add module sale_cutoff_time_delivery

grindtildeath committed:
fixup! fixup! fixup! fixup! fixup! fixup! Add module sale_cutoff_time_delivery

grindtildeath committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! Add module sale_cutoff_time_delivery

grindtildeath committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! Add module sale_cutoff_time_delivery

grindtildeath committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! Add module sale_cutoff_time_delivery

grindtildeath committed:
Add logging

jbaudoux commented on sale_cutoff_time_delivery/models/sale_order.py:
```diff
@@ -0,0 +1,123 @@
+            )
+        else:
+            utc_cutoff_time = cutoff_time
+        if expected_date.time() <= utc_cutoff_time:
```
You should not compare times. Apply here the same implementation you did in _prepare_procurement_values.
1:00 < 3:00 in CET so should remains the same after conversion to UTC: 23:00 <> 1:00

grindtildeath replied:
I'm not sure to understand why this wouldn't work since `utc_cutoff_time` is computed according to `expected_date`. Can you please explain a bit more?

LoisRForgeFlow replied:
@jbaudoux ping

jbaudoux replied:
Let's define a cutoff at 6AM AWST (Australia time). You order at 10AM, so you are after the cutoff.
Now I convert to UTC: Cutoff=22PM UTC. Order time is 2AM. I am before the cutoff.
If my warehouse is in Australia as well as my customers, I don't care about the time in London.

grindtildeath committed:
[IMP] sale_cutoff_time_delivery: black, isort

grindtildeath committed:
Readme

grindtildeath committed:
Add tests

jgrandguillaume reviewed (approved):
Functional test ok so far

jbaudoux commented:
@jgrandguillaume I don't approve yet. To discuss next week.

guewen commented on sale_cutoff_time_delivery/models/cutoff_time_mixin.py:
```diff
@@ -0,0 +1,40 @@
+
+class TimeCutoffMixin(models.AbstractModel):
+
+    _name = "time.cutoff.mixin"
```
```suggestion
    _name = "time.cutoff.mixin"
    _description = "Time Cut-off Mixin"
```

grindtildeath committed:
Update sale_cutoff_time_delivery/models/cutoff_time_mixin.py

Co-authored-by: Guewen Baconnier <guewen.baconnier@camptocamp.com>

simahawk commented:
@jbaudoux @grindtildeath any update on this?

grindtildeath commented:
@simahawk Not in my hands anymore :shrug:

Commit:
Apply cut-off time on commitment date

When a commitment date is set on the sale order, respect the date but
change the cut-off time. This will help to keep all scheduled datetimes
in line (e.g. if we want to merge them).

I added a lead time of 1 day in the test setup which is a usual configuration,
so I had to adapt all the other test results.

Commit:
Extract method to apply cut-off

Commit:
fixup! Apply cut-off time on commitment date

Commit:
Add OCA/partner-contact in oca_dependencies.txt

For sale_cutoff_time_delivery that needs partner_tz

simahawk commented:
replaced by #1378
